examples/rifle_sxiv.sh

```diff
@@ -10,39 +10,36 @@
 #
 #   mime ^image, has sxiv, X, flag f = path/to/this/script -- "$@"
 #
-# Implementation notes: this script is quite slow because of POSIX limitations
-# and portability concerns. First calling the shell function 'abspath' is
-# quicker than calling 'realpath' because it would fork a whole process, which
-# is slow. Second, we need to append a file list to sxiv, which can only be done
-# properly in two ways: arrays (which are not POSIX) or \0 sperated
-# strings. Unfortunately, assigning \0 to a variable is not POSIX either (will
-# not work in dash and others), so we cannot store the result of listfiles to a
-# variable.
 
-if [ $# -eq 0 ]; then
-    echo "Usage: ${0##*/} PICTURES"
-    exit
-fi
+tmp="/tmp/sxiv_rifle_$$"
 
-[ "$1" = '--' ] && shift
+listfiles () {
+    find -L "///${1%/*}" -maxdepth 1 -type f -iregex \
+      '.*\.\(jpe?g\|png\|gif\|webp\|tiff\|bmp\)$' -print | sort | tee "$tmp"
+}
 
-abspath () {
-    case "$1" in
-        /*) printf "%s\n" "$1";;
-        *)  printf "%s\n" "$PWD/$1";;
+is_img () {
+    case "${1##*.}" in
+        "jpg"|"jpeg"|"png"|"gif"|"webp"|"tiff"|"bmp") return 0 ;;
+        *) return 1 ;;
     esac
 }
 
-listfiles () {
-    find -L "$(dirname "$target")" -maxdepth 1 -type f -iregex \
-      '.*\(jpe?g\|bmp\|png\|gif\)$' -print0 | sort -z
+open_img () {
+    is_img "$1" || exit 1
+    trap 'rm -f $tmp' EXIT
+    count="$(listfiles "$1" | grep -nF "$1")"
```

Comment:
Hmm, won't dropping `-m1` cause way more of a potential slowdown?
If it's not POSIX I vote for doing the same thing I did with ripgrep.

```suggestion
    count="$(listfiles "$1" | grep -Fnm1 "$1" || grep -nF "$1")"
```

Reply:
> Hmm, won't dropping -m1 cause way more of a potential slowdown?

No, because unix pipes execute in parallel. This might be a valid concern while reading from a file.

- - -

Also that syntax doesn't really work. What's going to happen is this
```sh
cmd1 | cmd2 || cmd3
# is equal to
{ cmd1 | cmd2 } || cmd3
```

What we want is this
```sh
cmd1 | { cmd2 || cmd3 }
```
And even then this wont really work, because if cmd2 fails, `count` will store the failure/error msg as well.

If we want to add fallbacks we'll have to do it through storing the command in a variable. The `||` trick won't work.

Reply:
Doesn't the `-m1` still allow grep to short-circuit even though input is coming from stdin?

Reply:
Not really. Try this, `seq 50 | grep --invalid-flag || grep 1`.
First it'll print the error from invalid flag and then it'll just hang because the 2nd grep command didn't receive any input.

Reply:
Also, this would be the worst case scenario. 
```sh
[nrk ranger]% time seq 9000 | grep 1 >/dev/null
seq 9000  0.00s user 0.00s system 89% cpu 0.001 total
grep --color=auto 1 > /dev/null  0.00s user 0.00s system 81% cpu 0.001 total
[nrk ranger]% time seq 9000 | grep -m 1 1 >/dev/null
seq 9000  0.00s user 0.00s system 82% cpu 0.001 total
grep --color=auto -m 1 1 > /dev/null  0.00s user 0.00s system 82% cpu 0.001 total
```
I'm sending output to null since otherwise printing on the terminal becomes the bottleneck.

Reply:
The invalid flag output doesn't matter much. We could use the same solution as for ripgrep.

Fwiw, on my system I do see a 5x difference, more due to the number of matches though, seemingly. And ripgrep may not be worth the trouble, seems to have a slightly higher upfront execution cost and this script tends to run it on small input anyway.

fd might still prove worth the trouble.

Reply:
@toonn Sorry for leaving this hanging, was busy with a couple other things. So we're not going with ripgrep?

And for fd, I'm not familiar with it, but we stopped using null separated list. So assuming the first command was proper, simply changing `--print0` to `--print` should suffice, right?

```sh
fd --max-depth=1 --type f --ignore-case --absolute-path --print \
         -e jpg -e jpeg -e png -e webp -e tiff -e bmp "///${target%/*}" \
```

Reply:
No, just drop `--print0` altogether.

Dropping the null-separated paths is slightly unfortunate, since it'd allow generalizing the script for other things that *do* handle newlines fine. Not holding back the PR over that though.

Reply:
One last thing, does the directory need to be at then end of the args, or can it be at the beginning similar to `find` ?

Reply:
It can be before the flags but it has to be preceded by a pattern, use `.` for a pattern that matches anything. Feel free not to add it unless you can test the performance though. If there's no perceptible difference there's no need to complicate matters.
IME `fd` is a lot faster than `find` though.

Reply:
> Feel free not to add it unless you can test the performance though.

In that case let's drop it.